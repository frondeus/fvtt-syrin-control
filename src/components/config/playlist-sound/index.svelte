<script lang="ts">
	import Context from '@/services/context';
	import type { Soundsets } from '@/models/store';

	// Context
	const ctx = Context();

	// Params & State
	export let name: string;
	export let moodId: number;
	let soundsets = ctx.stores.soundsets;
	let soundsetName: string = '...';
	let moodName: string = '...';
	let style = '';
	let class_ = 'inner';

	const reactiveSoundsetName = async (moodId: number, soundsets: Soundsets) => {
		const soundset = await ctx.stores.hydrateSoundsetInner(
			(await ctx.api.soundsetIdForMood(moodId)) ?? '',
			soundsets
		);
		if (soundset === undefined) {
			return;
		}

		const mood = soundset.moods[moodId];

		moodName = mood.name;
		soundsetName = soundset.name;
		if (soundset.artworkUrl !== undefined) {
			style = `background-image: url('${soundset.artworkUrl}');`;
			class_ = 'inner inner-invert';
		}
	};

	$: reactiveSoundsetName(moodId, $soundsets);
</script>

<div class={class_} {style} data-test="syrin-playlist-sound-config">
	<div>
		<div class="form-group">
			<label for="">{ctx.game.localizeCore('PLAYLIST.SoundName')}</label>
			<input
				type="text"
				name="name"
				data-test="syrin-sound-name"
				placeholder={ctx.game.localizeCore('PLAYLIST.SoundName')}
				bind:value={name}
				required
			/>
		</div>
		<div class="form-group">
			<label for="">{ctx.game.localize('config.soundset')}</label>
			<input
				type="text"
				disabled
				data-test="syrin-soundset-name"
				value={soundsetName}
				title={soundsetName}
			/>
		</div>
		<div class="form-group">
			<label for="">{ctx.game.localize('config.mood')}</label>
			<input type="text" disabled data-test="syrin-mood-name" value={moodName} title={moodName} />
		</div>
	</div>

	<input type="hidden" data-test="syrin-path" name="path" value="syrinscape.wav" />

	<div>
		<div data-test="syrin-controlled">
			{ctx.game.localize('config.controlled', { name: ctx.game.localize('config.playlistSound') })}
		</div>
		<button type="submit">
			<i class="far fa-save" />
			{ctx.game.localizeCore('PLAYLIST.SoundUpdate')}
		</button>
	</div>
</div>

<style>
	.inner {
		padding: 8px;
		display: flex;
		flex-direction: column;
		justify-content: space-between;
		height: 100%;
	}
	.inner-invert {
		color: var(--color-text-light-highlight);
		background-size: cover;
	}
	.inner button,
	.inner input {
		background: url(/ui/parchment.jpg) repeat;
	}
</style>
