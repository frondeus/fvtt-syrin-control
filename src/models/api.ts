// What we get from Syrinscape
export interface ApiSoundset {
	id: number;
	uuid: string;
	name: string;
	full_name: string;
	artwork_url: string;
}

export interface ApiMood {
	pk: number;
	name: string;
	elements: ApiMoodElement[];
	soundset_uuid: string;
}

export interface ApiMoodElement {
	element: string; // Element url... last segment is pk
}

export interface ApiElement {
	name: string;
	pk: number;
	icon?: string;
	element_type: 'sfx' | 'oneshot';
}

export interface ApiStatus {
	global?: {
		soundset_uuid?: string;
	};
}
