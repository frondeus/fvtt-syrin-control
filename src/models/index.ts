export * from './api';
export * from './store';
