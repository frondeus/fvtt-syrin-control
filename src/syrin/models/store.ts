export interface Soundset {
	id: string;
	pid: number;
	name: string;
	moods: Moods;
	elements: Elements;
}

// export interface NextMood {
// 	moodId?: number;
// 	soundsetId?: number;
// }

export interface CurrentlyPlaying {
	mood: Mood;
	soundset: Soundset;
}

export interface Mood {
	id: number;
	name: string;
	soundset?: string;
	elementsIds: number[];
}

export interface Moods {
	[index: number]: Mood;
}

export interface Soundsets {
	[index: string]: Soundset;
}

export interface PlaylistEntry {
	mood: Mood;
	soundset: Soundset;
}

export interface Playlist {
	entries: PlaylistEntry[];
}

export interface Element {
	id: number;
	name: string;
	icon: string;
	type: "oneshot" | "sfx" | "music";
}

export interface Module {
	elementsApp: any;
}

export type Elements = Element[];

export interface GlobalElementsTab {
	kind: 'global';
}

export interface SoundsetElementsTab {
	kind: 'soundset';
	soundset: Soundset;
}

export type ElementsTab = SoundsetElementsTab | GlobalElementsTab;
export type ElementsTabs = ElementsTab[];

export interface PlaylistItem {
	isPlaying: boolean;
	mood: Mood;
	soundset: Soundset;
}

export interface AmbientSounds {
	[index: string]: AmbientSound
}

export type AmbientSound = MoodAmbientSound | ElementAmbientSound;

export interface MoodAmbientSound {
	kind: 'mood';
	volume: number;
	moodId: number;
}

export interface ElementAmbientSound {
	kind: 'element';
	volume: number;
	elementId: number;
}
