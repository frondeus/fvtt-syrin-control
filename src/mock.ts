import { RawApi } from '@/services/raw';
import { FVTTGame } from '@/services/game';
import { storeDefaults, Stores } from '@/services/stores';
import { Utils } from './services/utils';
import { Api } from './services/api';
import { Syrin } from './services/syrin';
import { Context } from './services/context';

export function mockedApi(): RawApi {
	return {
		stopMood: jest.fn(),
		playMood: jest.fn(),
		playElement: jest.fn(),
		getMoods: jest.fn(async (_) => []),
		getElements: jest.fn(async (_) => []),
		getSoundsets: jest.fn(async () => []),
		getGlobalElements: jest.fn(async () => []),
		onInit: jest.fn(),
		getState: jest.fn(),
		changePlayerVolume: jest.fn(),
		changeMoodVolume: jest.fn(),
		changeOneShotVolume: jest.fn(),
		playerJoined: jest.fn(),
		stopElement: jest.fn(),
		getMood: jest.fn()
	};
}

type StoreDefaultsKey = keyof typeof storeDefaults;
export function mockedGame(): FVTTGame {
	return {
		getSetting: jest.fn((name) => {
			if (Object.hasOwn(storeDefaults as object, name)) {
				return storeDefaults[name as StoreDefaultsKey] as any;
			}
			console.error('get setting', { name });
			return null!;
		}),
		setSetting: jest.fn(),
		setGlobal: jest.fn(),
		localize: jest.fn((key) => key),
		registerSetting: jest.fn(),
		notifyInfo: jest.fn(),
		notifyError: jest.fn(),
		callHookAll: jest.fn(),
		getActiveScene: jest.fn(),
		isGM: jest.fn(),
		setSettingToDefault: jest.fn(),
		socket: {
			register: jest.fn(),
			executeAsGM: jest.fn(),
			executeAsUser: jest.fn(),
			executeForAllGMs: jest.fn(),
			executeForEveryone: jest.fn(),
			executeForOtherGMs: jest.fn(),
			executeForOthers: jest.fn(),
			executeForUsers: jest.fn()
		},
		isReady: jest.fn(() => true),
		userId: jest.fn(),
		hasActiveModule: jest.fn(),
		localizeCore: jest.fn(),
		getAudioContext: jest.fn(),
		createMoodMacro: jest.fn(),
		createElementMacro: jest.fn(),
		createPlaylist: jest.fn(),
		createPlaylistSound: jest.fn(),
		createPlaylistMoodSound: jest.fn(),
		getPlaylists: jest.fn(),
		getAmbientSounds: jest.fn(),
		getPlayerName: jest.fn(),
		createDialog: jest.fn()
	};
}

export function mocked(game: FVTTGame = mockedGame(), raw: RawApi = mockedApi()): Context {
	const utils = new Utils(game);
	const api = new Api(raw);
	const stores = new Stores(game, api);
	const syrin = new Syrin(game, utils, api, stores);
	return new Context(game, utils, api, stores, syrin);
}
