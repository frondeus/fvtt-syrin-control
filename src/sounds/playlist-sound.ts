import { Context } from '@/services/context';
import { ElementSoundFlags, MoodSoundFlags } from '@/sounds';
import { Unsubscriber } from 'svelte/store';

export type SyrinPlaylistSoundFlags = ElementSoundFlags | MoodSoundFlags;

export interface PlaylistSoundProvider {
	playing(): boolean;
	id(): null | string;
	update(playing: boolean): void;
	ctx(): Context;
}

export class PlaylistSound {
	flags: SyrinPlaylistSoundFlags;
	unsubscriber?: Unsubscriber;
	wasPlaying: boolean;
	provider: PlaylistSoundProvider;

	constructor(data: any, provider: PlaylistSoundProvider) {
		this.flags = data.flags.syrinscape;
		this.wasPlaying = false;
		this.provider = provider;

		this.handleSubscribtion();
	}

	handleSubscribtion() {
		if (this.flags.type === 'mood' && this.provider.ctx().game.isGM()) {
			const moodId = this.flags.mood;
			this.unsubscriber = this.provider.ctx().stores.currentlyPlaying.subscribe((playing) => {
				const mood = playing?.mood;
				if (this.provider.id() !== null) {
					const playing = mood?.id === moodId;
					this.provider.update(playing);
					this.wasPlaying = playing;
				}
			});
		}
	}

	unsubscribe() {
		this.unsubscriber?.call([]);
	}

	async sync(): Promise<void> {
		if (!this.provider.ctx().api.isPlayerActive() || !this.provider.ctx().game.isGM()) {
			return;
		}

		if (this.provider.playing() !== this.wasPlaying) {
			switch (this.flags.type) {
				case 'mood': {
					if (this.provider.playing()) {
						this.provider.ctx().syrin.setMood(this.flags.mood);
					} else {
						this.provider.ctx().syrin.stopAll();
					}
					break;
				}
			}
		}
	}
}
