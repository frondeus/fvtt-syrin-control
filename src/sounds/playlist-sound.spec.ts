import { Mocked, mocked } from '@/mock';
import { ApiElement } from '@/models';
import { PlaylistSound, SyrinPlaylistSoundFlags } from './playlist-sound';

describe('Ambient sound controller', () => {
	let mock: Mocked;
	let flags: SyrinPlaylistSoundFlags;

	beforeEach(async () => {
		mock = mocked();
		mock.game.isGM = jest.fn(() => true);
		mock.game.userId = jest.fn(() => 'frondeus');
		mock.raw.getState = jest.fn(() => 'active');

		flags = {
			type: 'mood',
			mood: 4321
		};

		mock.raw.getMood = jest.fn(async (id) => {
			if (id === 4321) {
				return await import('@fixtures/raw-mood-4321.json');
			}
			return undefined;
		});
		mock.ctx.stores.soundsets.set({
			'1234': (await import('@fixtures/soundset-1234.json')).default
		});
		mock.raw.getElements = jest.fn(async (id) => {
			if (id === '1234') {
				return (await import('@fixtures/raw-elements-1234.json')).default as ApiElement[];
			}
			return [];
		});
	});

	it('calls unsubscriber', () => {
		const provider = {
			id: () => '1',
			playing: jest.fn(() => true),
			update: jest.fn((_playing: boolean) => {}),
			ctx: () => mock.ctx
		};
		const sound = new PlaylistSound({ flags: { syrinscape: flags } }, provider);
		const spy = jest.spyOn(sound, 'unsubscriber');

		sound.unsubscribe();

		expect(spy).toBeCalledTimes(1);
	});

	describe('subscribtion', () => {
		it('changes playing state when currentlyPlaying store is updated', async () => {
			let newPlaying = false;
			const provider = {
				id: () => '1',
				playing: jest.fn(() => false),
				update: jest.fn((playing: boolean) => {
					newPlaying = playing;
				}),
				ctx: () => mock.ctx
			};

			expect(newPlaying).toBe(false);

			const sound = new PlaylistSound({ flags: { syrinscape: flags } }, provider);

			mock.ctx.stores.nextPlaylistMood.set(4321);

			await new Promise((resolve) =>
				mock.ctx.stores.currentlyPlaying.subscribe((playing) => {
					if (playing !== undefined) resolve(playing);
				})
			);

			expect(newPlaying).toBe(true);

			sound.unsubscribe();
		});
	});

	describe('sync', () => {
		it('does nothing when player is inactive', async () => {
			const provider = {
				id: () => '1',
				playing: jest.fn(() => true),
				update: jest.fn((_playing: boolean) => {}),
				ctx: () => mock.ctx
			};
			const sound = new PlaylistSound({ flags: { syrinscape: flags } }, provider);

			mock.raw.getState = jest.fn(() => 'inactive');

			await sound.sync();

			expect(provider.playing).not.toBeCalled();
		});

		it('also does nothing when its not GM', async () => {
			mock.game.isGM = jest.fn(() => false);
			const provider = {
				id: () => '1',
				playing: jest.fn(() => true),
				update: jest.fn((_playing: boolean) => {}),
				ctx: () => mock.ctx
			};
			const sound = new PlaylistSound({ flags: { syrinscape: flags } }, provider);

			await sound.sync();

			expect(provider.playing).not.toBeCalled();
		});

		it('also does nothing when current playing state is no different from what we remember', async () => {
			const spySet = jest.spyOn(mock.ctx.syrin, 'setMood');
			const spyStop = jest.spyOn(mock.ctx.syrin, 'stopAll');
			const provider = {
				id: () => '1',
				playing: jest.fn(() => true),
				update: jest.fn((_playing: boolean) => {}),
				ctx: () => mock.ctx
			};
			const sound = new PlaylistSound({ flags: { syrinscape: flags } }, provider);
			sound.wasPlaying = true;

			await sound.sync();

			expect(spySet).not.toBeCalled();
			expect(spyStop).not.toBeCalled();
		});

		it('stops the mood', async () => {
			const spySet = jest.spyOn(mock.ctx.syrin, 'setMood');
			const spyStop = jest.spyOn(mock.ctx.syrin, 'stopAll');
			const provider = {
				id: () => '1',
				playing: jest.fn(() => false),
				update: jest.fn((_playing: boolean) => {}),
				ctx: () => mock.ctx
			};
			const sound = new PlaylistSound({ flags: { syrinscape: flags } }, provider);
			sound.wasPlaying = true;

			await sound.sync();

			expect(spyStop).toBeCalledTimes(1);

			expect(spySet).not.toBeCalled();
		});

		it('sets the mood', async () => {
			const spySet = jest.spyOn(mock.ctx.syrin, 'setMood');
			const spyStop = jest.spyOn(mock.ctx.syrin, 'stopAll');
			const provider = {
				id: () => '1',
				playing: jest.fn(() => true),
				update: jest.fn((_playing: boolean) => {}),
				ctx: () => mock.ctx
			};
			const sound = new PlaylistSound({ flags: { syrinscape: flags } }, provider);
			sound.wasPlaying = false;

			await sound.sync();

			expect(spySet).toBeCalledTimes(1);
			expect(spySet).toBeCalledWith(4321);

			expect(spyStop).not.toBeCalled();
		});
	});
	describe('bugs', () => {
		it('can be flattened', () => {
			const provider = {
				id: () => '1',
				playing: jest.fn(() => true),
				update: jest.fn((_playing: boolean) => {}),
				ctx: () => mock.ctx
			};
			const sound = new PlaylistSound({ flags: { syrinscape: flags } }, provider);

			const flattened = foundry.utils.flattenObject(sound);

			expect(flattened).toMatchSnapshot();
		});
	});
});
