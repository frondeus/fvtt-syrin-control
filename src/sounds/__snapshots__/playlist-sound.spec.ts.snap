// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ambient sound controller bugs can be flattened 1`] = `
{
  "flags.mood": 4321,
  "flags.type": "mood",
  "provider.ctx": [Function],
  "provider.id": [Function],
  "provider.playing": [MockFunction],
  "provider.update": [MockFunction] {
    "calls": [
      [
        false,
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "unsubscriber": [Function],
  "wasPlaying": false,
}
`;
